core/src/main/scala/org/http4s/Uri.scala

```diff
@@ -279,6 +275,126 @@ object Uri {
+    def addSegments(value: List[Path.Segment]): Path =
+      Path(this.segments ++ value, absolute = absolute || this.segments.isEmpty)
+
+    def normalize: Path = Path(segments.filterNot(_.isEmpty))
+
+    /* Merge paths per RFC 3986 5.2.3 */
+    def merge(path: Path): Path = {
+      val merge = segments.init
+      Path(merge ++ path.segments, absolute = absolute, endsWithSlash = path.endsWithSlash)
+    }
+
+    def startsWith(path: Path): Boolean = segments.startsWith(path.segments)
+
+    def startsWithString(path: String): Boolean = startsWith(Path.fromString(path))
+
+    def indexOf(path: Path): Option[Int] =
+      Some(segments.indexOfSlice(path.segments)).filterNot(_ == -1)
+
+    def indexOfString(path: String): Option[Int] = indexOf(Path.fromString(path))
+
+    def splitAt(idx: Int): (Path, Path) = {
+      val (start, end) = segments.splitAt(idx)
+      Path(start, absolute = absolute) -> Path(end, endsWithSlash = endsWithSlash)
+    }
+  }
+
+  object Path {
+    val empty = Path(Nil)
+    val Root = Path(Nil, absolute = true)
+
+    class Segment private (val encoded: String) {
+      def isEmpty = encoded.isEmpty
+
+      override def equals(obj: Any): Boolean = obj match {
+        case s: Segment => s.encoded == encoded
+      }
+
+      override def hashCode(): Int = encoded.hashCode
+
+      override val toString: String = encoded
+    }
+
+    object Segment extends (String => Segment) {
+      def apply(value: String): Segment = new Segment(pathEncode(value))
+      def encoded(value: String): Segment = new Segment(value)
+    }
+
+    /**
+      * This constructor allows you to construct the path directly.
+      * Each path segment needs to be encoded for it to be used here.
+      *
+      * @param segments the segments that this path consists of. MUST be Urlencoded.
+      * @param absolute if the path is absolute. I.E starts with a "/"
+      * @param endsWithSlash if the path is a "directory", ends with a "/"
+      * @return a Uri.Path that can be used in Uri, or by itself.
+      */
+    def apply(
+        segments: List[Segment],
+        absolute: Boolean = false,
+        endsWithSlash: Boolean = false): Path =
+      new Path(segments, absolute, endsWithSlash)
+
+    def unapply(path: Path): Some[(List[Segment], Boolean, Boolean)] =
+      Some((path.segments, path.absolute, path.endsWithSlash))
+
+    def fromString(fromPath: String): Path = fromPath match {
+      case "" => empty
+      case "/" => Root
+      case pth =>
+        val absolute = pth.startsWith("/")
+        val relative = if (absolute) pth.substring(1) else pth
+        Path(
+          segments = relative
+            .split("/")
+            .foldLeft(List.empty[Segment])((path, segment) =>
+              path ::: List(Segment.encoded(segment))),
+          absolute = absolute,
+          endsWithSlash = relative.endsWith("/")
+        )
+    }
+
+    implicit val eq: Eq[Path] = Eq.fromUniversalEquals[Path]
```

Comment:
A lot more are possible here.  `Hash`, `Order`.  I think `Monoid`.  `LowerBounded`, if you want to get exotic.  But it's more important to get this landed, and we can circle back to those.

Reply:
Not sure if this semigroup passes all the laws. I guess a test is in order

Reply:
I think you've even got a `Monoid` with `empty`.

We should decide if we want this to be an `Order` or not.  It could be.  I chose not to provide an `Order[Method]` in my recent reforms there, because I didn't know that alphabetical order was compelling.  It's perhaps a bit more useful to provide a stable ordering of URIs, of which this would be a part.

Reply:
There are two monoids, both `Uri.Path.empty` and `Uri.Path.Root` 
Although you might simulate the `Root` one with adding a `toAbsolute` after and vise versa with `toRelative`

Reply:
Oh. Hmm. I was thinking of `empty` as being the `empty`. Maybe we can sidestep the ambiguity by just providing `Semigroup`? I would need to think through the cases a bit more deeply.